File: nodejs/widget_control_traits.js
/**
 * Widget Control Traits — Node.js SDK Example
 *
 * Shows how to publish sensor readings with metadata_json to power the
 * IoT26 Widget Builder's UI control components.
 *
 * Widget UI Component → metadata_json field + values:
 *
 *   Toggle (Relay)  → {"mode": "on"} | {"mode": "off"}
 *   AC Mode         → {"mode": "cool"} | {"mode": "heat"} | {"mode": "off"}
 *   Door / Motor    → {"motor": "open"} | {"motor": "stop"} | {"motor": "close"}
 *   Brightness      → plain numeric value (no metadata needed)
 *   Display Text    → {"display": "Hello World"}
 *
 * The widget reads metadata_json from the latest sensor reading and highlights
 * the matching button. State persists across refreshes via the DB snapshot.
 *
 * Usage:
 *   IOT26_DEVICE_ID=<uuid> IOT26_DEVICE_TOKEN=<token> node widget_control_traits.js
 */

'use strict';

const { IoT26Client } = require('./index.js');

// ── Configuration ─────────────────────────────────────────────────────────────

const DEVICE_ID    = process.env.IOT26_DEVICE_ID    || 'your-device-uuid';
const DEVICE_TOKEN = process.env.IOT26_DEVICE_TOKEN || 'your-device-token';
const BROKER       = process.env.IOT26_BROKER       || 'mqtt://localhost:1883';

// ── Sensor IDs — replace with UUIDs from your IoT26 sensor config ─────────────

const SENSOR_RELAY      = 'relay-sensor-uuid';      // Toggle widget (on/off)
const SENSOR_AC_MODE    = 'ac-mode-sensor-uuid';    // AC widget    (cool/heat/off)
const SENSOR_DOOR_MOTOR = 'door-motor-sensor-uuid'; // Door widget  (open/stop/close)
const SENSOR_BRIGHTNESS = 'brightness-sensor-uuid'; // Slider widget (0–100%)
const SENSOR_DISPLAY    = 'display-sensor-uuid';    // Display text widget

// ── Device state (simulated — replace with real hardware reads) ───────────────

const state = {
  relay:      'off',    // 'on' | 'off'
  acMode:     'off',    // 'cool' | 'heat' | 'off'
  doorMotor:  'close',  // 'open' | 'stop' | 'close'
  brightness: 75,       // 0–100
  displayMsg: 'Hello!',
};

// ── Helpers ───────────────────────────────────────────────────────────────────

function buildReadings() {
  return [
    // Toggle/Relay: mode = "on" or "off"
    {
      sensorId: SENSOR_RELAY,
      value:    state.relay === 'on' ? 1.0 : 0.0,
      unit:     'state',
      metadataJson: JSON.stringify({ mode: state.relay }),
    },

    // AC Mode: mode = "cool" | "heat" | "off"
    {
      sensorId:     SENSOR_AC_MODE,
      value:        1.0,
      unit:         'state',
      metadataJson: JSON.stringify({ mode: state.acMode }),
    },

    // Door / Motor: motor = "open" | "stop" | "close"
    {
      sensorId:     SENSOR_DOOR_MOTOR,
      value:        1.0,
      unit:         'state',
      metadataJson: JSON.stringify({ motor: state.doorMotor }),
    },

    // Brightness: plain numeric value — slider reads the value directly
    {
      sensorId: SENSOR_BRIGHTNESS,
      value:    state.brightness,
      unit:     '%',
    },

    // Display text: display = "any text"
    {
      sensorId:     SENSOR_DISPLAY,
      value:        1.0,
      unit:         'text',
      metadataJson: JSON.stringify({ display: state.displayMsg }),
    },
  ];
}

async function publishAll(client) {
  await client.publishReadings(buildReadings());
  console.log(`↑ State pushed  relay=${state.relay} ac=${state.acMode} door=${state.doorMotor} brightness=${state.brightness} display="${state.displayMsg}"`);
}

// ── Command handler ───────────────────────────────────────────────────────────

function makeCommandHandler(client) {
  return async (cmd) => {
    console.log(`↓ Command received: action=${cmd.action}`);

    if (cmd.action !== 'custom') {
      console.log(`  Ignoring non-custom action: ${cmd.action}`);
      return;
    }

    const custom = (cmd.raw && cmd.raw.custom) ? cmd.raw.custom : {};

    // ── Relay ──────────────────────────────────────────────────────────────
    if (custom.relay !== undefined) {
      state.relay = custom.relay ? 'on' : 'off';
      console.log(`  Relay → ${state.relay}`);
      await client.publishReadings([{
        sensorId:     SENSOR_RELAY,
        value:        state.relay === 'on' ? 1.0 : 0.0,
        unit:         'state',
        metadataJson: JSON.stringify({ mode: state.relay }),
      }]);
    }

    // ── AC Mode ────────────────────────────────────────────────────────────
    if (custom.ac_mode !== undefined) {
      const validModes = ['cool', 'heat', 'off'];
      if (validModes.includes(custom.ac_mode)) {
        state.acMode = custom.ac_mode;
        console.log(`  AC mode → ${state.acMode}`);
        await client.publishReadings([{
          sensorId:     SENSOR_AC_MODE,
          value:        1.0,
          unit:         'state',
          metadataJson: JSON.stringify({ mode: state.acMode }),
        }]);
      }
    }

    // ── Door / Motor ───────────────────────────────────────────────────────
    if (custom.motor !== undefined) {
      const validPos = ['open', 'stop', 'close'];
      if (validPos.includes(custom.motor)) {
        state.doorMotor = custom.motor;
        console.log(`  Door motor → ${state.doorMotor}`);
        await client.publishReadings([{
          sensorId:     SENSOR_DOOR_MOTOR,
          value:        1.0,
          unit:         'state',
          metadataJson: JSON.stringify({ motor: state.doorMotor }),
        }]);
      }
    }

    // ── Brightness ─────────────────────────────────────────────────────────
    if (custom.brightness !== undefined) {
      state.brightness = parseFloat(custom.brightness);
      console.log(`  Brightness → ${state.brightness}%`);
      await client.publishReadings([{
        sensorId: SENSOR_BRIGHTNESS,
        value:    state.brightness,
        unit:     '%',
      }]);
    }

    // ── Display text ───────────────────────────────────────────────────────
    if (custom.display !== undefined) {
      state.displayMsg = String(custom.display);
      console.log(`  Display → "${state.displayMsg}"`);
      await client.publishReadings([{
        sensorId:     SENSOR_DISPLAY,
        value:        1.0,
        unit:         'text',
        metadataJson: JSON.stringify({ display: state.displayMsg }),
      }]);
    }
  };
}

// ── Main ──────────────────────────────────────────────────────────────────────

async function main() {
  const client = new IoT26Client({
    deviceId:    DEVICE_ID,
    deviceToken: DEVICE_TOKEN,
    broker:      BROKER,
  });

  client.onCommand(makeCommandHandler(client));

  await client.connect();
  console.log('✓ Connected to IoT26 MQTT broker');

  // Push initial state so the widget immediately shows correct button highlights
  console.log('Publishing initial device state...');
  await publishAll(client);

  console.log('Running — widget reflects state in real time. Press Ctrl+C to stop.');

  // Re-publish heartbeat state every 30 s
  setInterval(() => publishAll(client), 30_000);
}

main().catch((err) => {
  console.error('Fatal error:', err);
  process.exit(1);
});

File: python/examples/widget_control_traits.py
"""
Widget Control Traits — Python SDK Example
==========================================

This example demonstrates how to publish sensor readings with `metadata_json`
to power the IoT26 Widget Builder's UI control components.

The Widget Builder supports the following UI components that derive their
displayed state from sensor readings published by your device:

  ┌─────────────────┬────────────────────────────────────────────────────┐
  │ UI Component    │ metadata_json field + values                       │
  ├─────────────────┼────────────────────────────────────────────────────┤
  │ Toggle (Relay)  │ {"mode": "on"} or {"mode": "off"}                  │
  │ AC Mode         │ {"mode": "cool"} or {"mode": "heat"} or {"mode": "off"} │
  │ Door / Motor    │ {"motor": "open"} or {"motor": "stop"} or {"motor": "close"} │
  │ Brightness      │ plain value (e.g. 80.0) — no metadata needed       │
  │ Display Text    │ {"display": "Hello World"}                         │
  └─────────────────┴────────────────────────────────────────────────────┘

The Widget reads `metadata_json` from the latest sensor reading and highlights
the matching button. The state also persists across page refreshes (fetched
from the database on reconnect).

Usage:
    pip install iot26-edge
    IOT26_DEVICE_ID=<uuid> IOT26_DEVICE_TOKEN=<token> python widget_control_traits.py

Environment variables:
    IOT26_DEVICE_ID     — Device UUID from IoT26 dashboard
    IOT26_DEVICE_TOKEN  — Device JWT token from IoT26 dashboard
    IOT26_BROKER        — MQTT broker host (default: localhost)
    IOT26_API_BASE      — REST API base URL (default: http://localhost:8443)
"""

import json
import logging
import os
import time
from iot26_edge.client import IoT26Client

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("widget_control_traits")

# ── Configuration ────────────────────────────────────────────────────────────

DEVICE_ID    = os.environ.get("IOT26_DEVICE_ID",    "your-device-uuid")
DEVICE_TOKEN = os.environ.get("IOT26_DEVICE_TOKEN", "your-device-token")
MQTT_BROKER  = os.environ.get("IOT26_BROKER",       "localhost")
API_BASE     = os.environ.get("IOT26_API_BASE",      "http://localhost:8443")

# ── Sensor IDs — replace with UUIDs from your IoT26 sensor config ───────────

SENSOR_RELAY      = "relay-sensor-uuid"     # Toggle widget (on/off)
SENSOR_AC_MODE    = "ac-mode-sensor-uuid"   # AC widget    (cool/heat/off)
SENSOR_DOOR_MOTOR = "door-motor-sensor-uuid"# Door widget  (open/stop/close)
SENSOR_BRIGHTNESS = "brightness-sensor-uuid"# Slider widget (0–100%)
SENSOR_DISPLAY    = "display-sensor-uuid"   # Display text widget


# ── Device state (simulated — replace with real hardware reads) ──────────────

state = {
    "relay":      "off",      # "on" | "off"
    "ac_mode":    "off",      # "cool" | "heat" | "off"
    "door_motor": "close",    # "open" | "stop" | "close"
    "brightness": 75.0,       # 0–100
    "display":    "Hello!",   # arbitrary string
}


def publish_all_states(client: IoT26Client) -> None:
    """Push the current device state to IoT26 so the widget reflects reality."""
    client.publish_batch([
        # Toggle/Relay: mode = "on" or "off"
        {
            "sensor_id": SENSOR_RELAY,
            "value":     1.0 if state["relay"] == "on" else 0.0,
            "unit":      "state",
            "metadata_json": json.dumps({"mode": state["relay"]}),
        },

        # AC Mode: mode = "cool" | "heat" | "off"
        {
            "sensor_id": SENSOR_AC_MODE,
            "value":     1.0,
            "unit":      "state",
            "metadata_json": json.dumps({"mode": state["ac_mode"]}),
        },

        # Door / Motor: motor = "open" | "stop" | "close"
        {
            "sensor_id": SENSOR_DOOR_MOTOR,
            "value":     1.0,
            "unit":      "state",
            "metadata_json": json.dumps({"motor": state["door_motor"]}),
        },

        # Brightness: plain numeric value — slider reads the value directly
        {
            "sensor_id": SENSOR_BRIGHTNESS,
            "value":     state["brightness"],
            "unit":      "%",
        },

        # Display text: display = "any text"
        {
            "sensor_id": SENSOR_DISPLAY,
            "value":     1.0,
            "unit":      "text",
            "metadata_json": json.dumps({"display": state["display"]}),
        },
    ])
    log.info("Pushed state → relay=%s, ac=%s, door=%s, brightness=%.0f",
             state["relay"], state["ac_mode"], state["door_motor"], state["brightness"])


def handle_command(cmd: dict) -> None:
    """
    Process downlink commands from the widget and publish uplink
    confirmation so the UI immediately reflects the new state.
    """
    action = cmd.get("action")
    custom = cmd.get("custom", {})
    log.info("⬇ Command received: action=%s  custom=%s", action, custom)

    if action != "custom":
        log.info("Ignoring non-custom action: %s", action)
        return

    # ── Relay toggle ─────────────────────────────────────────────────────────
    if "relay" in custom:
        state["relay"] = "on" if custom["relay"] else "off"
        log.info("Relay → %s", state["relay"])
        client.publish_batch([{
            "sensor_id": SENSOR_RELAY,
            "value":     1.0 if state["relay"] == "on" else 0.0,
            "unit":      "state",
            "metadata_json": json.dumps({"mode": state["relay"]}),
        }])

    # ── AC mode ──────────────────────────────────────────────────────────────
    if "ac_mode" in custom:
        new_mode = custom["ac_mode"]            # "cool" | "heat" | "off"
        if new_mode in ("cool", "heat", "off"):
            state["ac_mode"] = new_mode
            log.info("AC mode → %s", state["ac_mode"])
            client.publish_batch([{
                "sensor_id": SENSOR_AC_MODE,
                "value":     1.0,
                "unit":      "state",
                "metadata_json": json.dumps({"mode": state["ac_mode"]}),
            }])

    # ── Door / Motor ─────────────────────────────────────────────────────────
    if "motor" in custom:
        new_pos = custom["motor"]               # "open" | "stop" | "close"
        if new_pos in ("open", "stop", "close"):
            state["door_motor"] = new_pos
            log.info("Door motor → %s", state["door_motor"])
            client.publish_batch([{
                "sensor_id": SENSOR_DOOR_MOTOR,
                "value":     1.0,
                "unit":      "state",
                "metadata_json": json.dumps({"motor": state["door_motor"]}),
            }])

    # ── Brightness slider ─────────────────────────────────────────────────────
    if "brightness" in custom:
        state["brightness"] = float(custom["brightness"])
        log.info("Brightness → %.0f%%", state["brightness"])
        client.publish_batch([{
            "sensor_id": SENSOR_BRIGHTNESS,
            "value":     state["brightness"],
            "unit":      "%",
        }])

    # ── Display text ──────────────────────────────────────────────────────────
    if "display" in custom:
        state["display"] = str(custom["display"])
        log.info("Display → %r", state["display"])
        client.publish_batch([{
            "sensor_id": SENSOR_DISPLAY,
            "value":     1.0,
            "unit":      "text",
            "metadata_json": json.dumps({"display": state["display"]}),
        }])


# ── Main ─────────────────────────────────────────────────────────────────────

client = IoT26Client(
    device_id=DEVICE_ID,
    device_token=DEVICE_TOKEN,
    api_base=API_BASE,
    mqtt_broker=MQTT_BROKER,
    mqtt_port=1883,
)

client.on_command(handle_command)
client.connect_mqtt()

# Give the MQTT loop a moment to connect
time.sleep(1.5)

# Push initial state so the widget immediately shows correct button highlights
log.info("Publishing initial device state...")
publish_all_states(client)

try:
    log.info("Running — widget will reflect state in real time. Press Ctrl+C to stop.")
    while True:
        time.sleep(30)
        # Optionally re-publish heartbeat state every 30 s
        publish_all_states(client)

except KeyboardInterrupt:
    log.info("Shutting down...")
finally:
    client.disconnect()
